bittensor/_dendrite/dendrite_impl.py

```diff
@@ -351,6 +351,121 @@ def _forward(
+                If the model should early stop if the probabilty drops a certain threshold
+            num_return_sequences: (:obj: int, :default: 1):
+                How many sequences should the model return
+            do_sample (:obj: bool, :default: False):
+                If the model should do sample its probablity during generation
+            top_p (:obj: float, :default: 0.95): 
+                probability cutoff for top p sampling
+            temperature: (:obj: float, :default: 1.0):
+                The value used to module the next token probabilities for the softmax calculation
+            repetition_penalty (:obj: float, :default: 1.0):
+                The parameter for repetition penalty. 1.0 means no penalty.
+            length_penalty (:obj: float, :default: 1.0): 
+                The parameter for length penalty. 0.0 means no penalty, <0 to encourage longer sequences.
+            max_time (:obj: float, :default: 150): 
+                The maximum time that a server can use to generate
+            num_beam_groups (:obj: int, :default: 1):
+                Number of groups to divide num_beams into in order to ensure diversity among different groups of beams. 
+        Returns:
+            codes (:obj:`List[str]`, `required`):
+                Parsed codes from each endpoint from query.
+
+            times (:obj:`List[float]`, `required`):
+                Query times for each call from each endpoint.
+
+            generations (:obj:`List[str]`, `required`):
+                Generations from each endpoint.
+        """
+        tokenizer = bittensor.tokenizer()
+        response = self.text ( 
+            endpoints=endpoints, 
+            inputs = prompt,
+            synapses = [
+                synapse.TextSeq2Seq(
+                    topk = topk,
+                    num_to_generate = num_to_generate,
+                    num_beams = num_beams,
+                    no_repeat_ngram_size = no_repeat_ngram_size,
+                    early_stopping = early_stopping,
+                    num_return_sequences = num_return_sequences,
+                    do_sample = do_sample,
+                    top_p = top_p,
+                    temperature = temperature,
+                    repetition_penalty = repetition_penalty,
+                    length_penalty = length_penalty,
+                    max_time = max_time,
+                    num_beam_groups = num_beam_groups,
+                )
+            ],
+            timeout = timeout
+        )
+        # Parse responses to natural language.
+        generations = []
+        for text_tensor in response[0]:
+            generations.append( tokenizer.decode( text_tensor[0][0].long() ) )
+        codes = []
+        for code_tensor in response[1]:
+            codes.append( bittensor.utils.codes.code_to_string( code_tensor ) )
+        times = []
+        for time_tensor in response[2]:
+            times.append( time_tensor.item() )
```

Comment:
Every single for is a nice candidate to abstract into a new separated function and test:
- You get the correct index: 0, 1 or 2 depending on the function
- All elements were appended as expected
- you called the needed external function (that will be mocked): tokenizer.decode / bittensor.utils.codes.code_to_string / time_tensor.item